[androiddebugbridge] initial contribution

GiviMAD opened the pull request:
Signed-off-by: Miguel <miguelwork92@gmail.com>

Hi openHAB community. I have created this binding to allow openHAB to connect to android devices using the adb protocol.
My motivation for this was to replace some adb commands that I was sending using the exec-binding (to an android device connected to my tv). With the binding I can also avoid to install de adb cli in the docker image and use the official image directly.
I also added some other channels like a player, a volume dimmer, and some to control the application opened. I've been using it for a couple of days and it seems to work correctly. 
Hope it can be helpful for others.

cweitkamp commented:
Thank you very much for your contribution. Will it be possible to control / connect to Fire TV Sticks by this binding? I would be very much interested in that feature.

GiviMAD commented:
@cweitkamp. I'm happy to say it does :D , at least the Firestick 4k is the device that I'm using with the binding right now. Everything seems to work nicely unless the audio dimmer which I was not able to read/write from the adb. The audio dimmer works properly on the other device that I test with my old nexus 5x.

GiviMAD commented:
I can't find the build for this pr on the ci. I think there was some kind of error and it was not created. Can you trigger a rebuild? Thanks in advance.

cweitkamp commented:
> Can you trigger a rebuild? Thanks in advance.

Yes, of course. You can do that on your own by adding the "rebuild" label to your PR.

fwolter commented on bundles/org.openhab.binding.androiddebugbridge/pom.xml:
```diff
@@ -0,0 +1,32 @@
+
+  <name>openHAB Add-ons :: Bundles :: AndroidDebugBridge Binding</name>
+
+  <repositories>
+    <repository>
+      <id>jitpack.io</id>
+      <url>https://jitpack.io</url>
+    </repository>
+  </repositories>
```
Can you ask the developer to make it available on a repository openHAB supports?

GiviMAD replied:
I find out that there is a fork of the project more maintained and available on maven. I will move to that one.

fwolter commented on bundles/org.openhab.binding.androiddebugbridge/src/main/feature/feature.xml:
```diff
@@ -0,0 +1,23 @@
+<features name="org.openhab.binding.androiddebugbridge-${project.version}" xmlns="http://karaf.apache.org/xmlns/features/v1.4.0">
+	<repository>mvn:org.openhab.core.features.karaf/org.openhab.core.features.karaf.openhab-core/${ohc.version}/xml/features</repository>
+
+	<feature name="openhab-binding-androiddebugbridge" description="AndroidDebugBridge Binding" version="${project.version}">
```
```suggestion
	<feature name="openhab-binding-androiddebugbridge" description="Android Debug Bridge Binding" version="${project.version}">
```

fwolter commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeDevice.java:
```diff
@@ -0,0 +1,329 @@
+    private static final Pattern VOLUME_PATTERN = Pattern
+            .compile("volume is (?<current>\\d.*) in range \\[(?<min>\\d.*)\\.\\.(?<max>\\d.*)]");
+
+    @Nullable
+    static AdbCrypto adbCrypto;
```
Is this field package-private by intention by not specifying `private` or `public`?
```suggestion
    static @Nullable AdbCrypto ADB_CRYPTO;
```

fwolter commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeDevice.java:
```diff
@@ -0,0 +1,329 @@
+    static {
+        var logger = LoggerFactory.getLogger(AndroidDebugBridgeDevice.class);
+        try {
+            File directory = new File(".adb");
+            if (!directory.exists()) {
+                directory.mkdir();
+            }
+            adbCrypto = loadKeyPair(".adb/adb_pub.key", ".adb/adb.key");
```
You could create these files in OH's userdata directory. There are constants available to get the path.

fwolter commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeDevice.java:
```diff
@@ -0,0 +1,329 @@
+    @Nullable
+    private AdbConnection connection;
+
+    public AndroidDebugBridgeDevice() {
+    }
```
Can this be removed?

fwolter commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeDevice.java:
```diff
@@ -0,0 +1,329 @@
+        throw new AndroidDebugBridgeDeviceReadException("can read package name");
+    }
+
+    // public String[] getPackages(String packageName)
+    // throws IOException, AndroidDebugBridgeDeviceNotConnectedException, InterruptedException {
+    // var out = runAdbShell("cmd", "package", "list", "packages");
+    // return Arrays.asList(out.split("\n")).stream().filter(line -> line.contains("package:"))
+    // .map(line -> line.substring("package:".length())).toArray(String[]::new);
+    // }
```
Can this be removed?

fwolter commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeDevice.java:
```diff
@@ -0,0 +1,329 @@
+        runAdbShell("am", "force-stop", packageName);
+    }
+
+    public String getCurrentPackage() throws IOException, AndroidDebugBridgeDeviceNotConnectedException,
+            InterruptedException, AndroidDebugBridgeDeviceException, AndroidDebugBridgeDeviceReadException {
```
`AndroidDebugBridgeDeviceException` is never thrown.

fwolter commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeDevice.java:
```diff
@@ -0,0 +1,329 @@
+
+    private VolumeInfo getVolume(int stream) throws IOException, AndroidDebugBridgeDeviceNotConnectedException,
+            InterruptedException, AndroidDebugBridgeDeviceReadException {
+        String volumeResp = runAdbShell("media", "volume", "--show", "--stream", String.valueOf(stream), "--get", "|",
```
```suggestion
        String volumeResp = runAdbShell("media", "volume", "--show", "--stream", stream, "--get", "|",
```

GiviMAD replied:
Seems like I can't apply this change. runAdbShell expect this "String... args".

fwolter replied:
You're right, that won't work.

fwolter commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeDevice.java:
```diff
@@ -0,0 +1,329 @@
+        try {
+            sock = new Socket(ip, port);
+            sock.setTcpNoDelay(true);
+            // sock.setSoTimeout(5000);
```
Can this be removed?

fwolter commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeDiscoveryService.java:
```diff
@@ -0,0 +1,171 @@
+                                    break;
+                                }
+                            }
+                        } catch (IOException | AndroidDebugBridgeDevice.AndroidDebugBridgeDeviceException
+                                | AndroidDebugBridgeDevice.AndroidDebugBridgeDeviceCryptographyException
+                                | AndroidDebugBridgeDevice.AndroidDebugBridgeDeviceNotConnectedException
+                                | AndroidDebugBridgeDevice.AndroidDebugBridgeDeviceReadException e) {
```
If you don't need to distinguish between the exception types, you could use the same type with a different message.

fwolter commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeDiscoveryService.java:
```diff
@@ -0,0 +1,171 @@
+    private void onDiscoverResult(String serialNo, String ip, int port, String model, String androidVersion,
+            String brand) {
+        Map<String, Object> properties = new HashMap<>();
+        properties.put(PARAMETER_SERIAL, serialNo);
+        properties.put(PARAMETER_IP, ip);
+        properties.put(PARAMETER_PORT, port);
+        properties.put(PARAMETER_MODEL, model);
+        properties.put(PARAMETER_BRAND, brand);
+        properties.put(PARAMETER_ANDROID_VERSION, androidVersion);
```
Please use `Thing.PROPERTY_*` where applicable.

fwolter commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeDiscoveryService.java:
```diff
@@ -0,0 +1,171 @@
+        properties.put(PARAMETER_ANDROID_VERSION, androidVersion);
+        thingDiscovered(DiscoveryResultBuilder.create(new ThingUID(THING_TYPE_ANDROID_DEVICE, serialNo))
+                .withTTL(DISCOVERY_RESULT_TTL_SEC).withRepresentationProperty(PARAMETER_SERIAL)
+                .withProperties(properties).withLabel(String.format("%s(%s)", model, serialNo)).build());
```
```suggestion
                .withProperties(properties).withLabel(String.format("%s (%s)", model, serialNo)).build());
```

fwolter commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeHandler.java:
```diff
@@ -0,0 +1,247 @@
+            schedule.cancel(true);
+            connectionCheckerSchedule = null;
+        }
+        adbConnection.disconnect();
```
How long can this take at maximum? `dispose()` is supposed to return fast.

GiviMAD replied:
I made some test and it takes less than 5 ms always.

GiviMAD replied:
I've never face that problem when I was testing the device disconnection and reconnection. Also I have reviewed the implementation and I don't see any reason why its execution could be blocked by a network problem.

fwolter commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeHandler.java:
```diff
@@ -0,0 +1,247 @@
+                    refreshStatus();
+                }
+            }
+        } catch (Exception e) {
+            // retry on unexpected errors
```
Unchecked exceptions are handled by the framework. So, this could be removed.

fwolter commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeHandlerFactory.java:
```diff
@@ -0,0 +1,107 @@
+        return null;
+    }
+
+    private synchronized void registerDiscoveryService(AndroidDebugBridgeBindingConfiguration config) {
+        AndroidDebugBridgeDiscoveryService androidADBDiscoveryService = discoveryService;
+        if (androidADBDiscoveryService != null) {
+            androidADBDiscoveryService.updateConfig(config);
+        } else {
+            androidADBDiscoveryService = new AndroidDebugBridgeDiscoveryService();
+            androidADBDiscoveryService.updateConfig(config);
+            discoveryService = androidADBDiscoveryService;
+            discoveryServiceRegistration = bundleContext.registerService(DiscoveryService.class.getName(),
+                    androidADBDiscoveryService, null);
+        }
+    }
+
+    private void unregisterDiscoveryService() {
+        var adbDiscoveryServiceRegistration = discoveryServiceRegistration;
+        if (adbDiscoveryServiceRegistration != null) {
+            adbDiscoveryServiceRegistration.unregister();
+            discoveryServiceRegistration = null;
+        }
+    }
+
+    @Activate
+    protected void activate(ComponentContext componentContext, Map<String, Object> config) {
+        super.activate(componentContext);
+        modified(config);
+    }
```
You could use `ThingHandlerService` which would make this code unnecessary. The documentation for this is not yet merged, but available in the PR: https://github.com/openhab/openhab-docs/pull/1262/files#diff-c4a4d8725430bc2ea046182bfc73ac51349d989c7f6a8a6fa3001a226a09ad98R932

GiviMAD replied:
I was reviewing the documentation in the pr and seems like I need to create a bridge, right now I'm just passing the binding configurations to the discovery. 
Is there any other way to access to the binding configuration from the discovery service? 
Or should I create a bridge just to configure the service discovery?
Maybe I'm missing something.
Thanks in advance.

fwolter replied:
If your discovery implements `ThingHandlerService` you need to implement the interface method `setThingHandler(ThingHandler handler)`. Then, you can retrieve the bridge's config from the handler: `handler.getThing().getConfiguration()`

GiviMAD replied:
It doesn't seem to fit my case because I don't have any bridge implemented on the binding. I configure the discovery through the binding configuration.

fwolter replied:
Now I get it. The binding configuration is stored as properties in the `AndroidDebugBridgeHandlerFactory` service. I never done this, but if you're keen, you could retrieve the config by injecting `ConfigurationAdmin` in the discovery class:

```
    @Reference
    void setConfigurationAdmin(ConfigurationAdmin admin) {
        this.admin = admin;
    }
 
    public void ranking() throws IOException {
        Configuration configOnline =
            this.admin.getConfiguration(
                "org.fipro.ds.data.online.OnlineDataService",
                null);
        Dictionary<String, Object> propsOnline = null;
        if (configOnline != null
                && configOnline.getProperties() != null) {
            propsOnline = configOnline.getProperties();
        }
```

The string argument to `admin.getConfiguration()` must be the PID of your factory: "binding.androiddebugbridge"

Then, you should be able to obtain a config parameter by calling for example `propsOnline.get("discoveryPort")`. I'm curious if this works.

fwolter commented on bundles/org.openhab.binding.androiddebugbridge/src/main/resources/OH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,32 @@
+
+	<name>AndroidDebugBridge Binding</name>
+	<description>This is the binding for connect to Android devices using the Android Debug Bridge protocol.</description>
+	<author>Miguel Álvarez</author>
```
The author tag is deprecated and should therefore be removed. See openhab/openhab-core#1844.

fwolter commented on bundles/org.openhab.binding.androiddebugbridge/src/main/resources/OH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,32 @@
+	xmlns:binding="https://openhab.org/schemas/binding/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/binding/v1.0.0 https://openhab.org/schemas/binding-1.0.0.xsd">
+
+	<name>AndroidDebugBridge Binding</name>
```
```suggestion
	<name>Android Debug Bridge Binding</name>
```

fwolter commented on bundles/org.openhab.binding.androiddebugbridge/src/main/resources/OH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,32 @@
+			<description>Port used on discovery to connect to the device through adb.</description>
+			<default>5555</default>
+		</parameter>
+		<parameter name="discoveryReachableMs" type="integer" required="true">
```
```suggestion
		<parameter name="discoveryReachableMs" type="integer" required="true" unit="ms">
```

fwolter commented on bundles/org.openhab.binding.androiddebugbridge/src/main/resources/OH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,32 @@
+			<default>5555</default>
+		</parameter>
+		<parameter name="discoveryReachableMs" type="integer" required="true">
+			<label>Discovery Reachable ms</label>
```
Labels are expected to be as short as possible. Guideline is 2-3 words with up to 25 chars. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions
```suggestion
			<label>Discovery Reachable</label>
```

fwolter commented on bundles/org.openhab.binding.androiddebugbridge/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,375 @@
+				<label>ADB Port</label>
+				<description>Device port listening to adb connections.</description>
+			</parameter>
+			<parameter name="refreshTimeSec" type="integer" min="10" max="120" required="true">
```
```suggestion
			<parameter name="refreshTimeSec" type="integer" min="10" max="120" unit="s" required="true">
```

fwolter commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeHandler.java:
```diff
@@ -0,0 +1,247 @@
+            var currentConfig = config;
+            if (currentConfig == null)
+                return;
+            logger.debug("Refresh device {} status", config.ip);
```
```suggestion
            logger.debug("Refresh device {} status", currentConfig.ip);
```

fwolter reviewed (changes requested):
Thanks for your contribution! I reviewed your code and here is my feedback.

GiviMAD commented on bundles/org.openhab.binding.androiddebugbridge/README.md:
```diff
@@ -0,0 +1,63 @@
+
+| channel  | type   | description                  |
+|----------|--------|------------------------------|
+| key-event  | String | Send key event to android device |
```
I forgot to add the possible values. These are defined on the thing description but I will add them also here. Thanks.

GiviMAD commented:
I think everything is done. 
Also the build fails to queue again and I can't edit the pr labels, if no more changes are required can you trigger the build again? Thanks!

fwolter commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeDiscoveryService.java:
```diff
@@ -0,0 +1,192 @@
+                .withProperties(properties).withLabel(String.format("%s (%s)", model, serialNo)).build());
+    }
+
+    @Reference
+    void setConfigurationAdmin(ConfigurationAdmin admin) {
+        this.admin = admin;
+    }
```
Great to see this works!

You could inject this via the constructor to get rid of the Nullable annotation.

fwolter commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeDiscoveryService.java:
```diff
@@ -0,0 +1,192 @@
+            return null;
+        }
+        try {
+            Configuration configOnline = currentAdmin.getConfiguration("binding.androiddebugbridge", null);
```
Can you make "binding.androiddebugbridge" a constant?

fwolter commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeHandlerFactory.java:
```diff
@@ -0,0 +1,49 @@
+ * @author Miguel Álvarez - Initial contribution
+ */
+@NonNullByDefault
+@Component(configurationPid = "binding.androiddebugbridge", service = ThingHandlerFactory.class)
```
I think you could use the constant here, too.

fwolter commented on bundles/org.openhab.binding.androiddebugbridge/README.md:
```diff
@@ -0,0 +1,62 @@
+
+| channel  | type   | description                  |
+|----------|--------|------------------------------|
+| key-event  | String | Send key event to android device. Possible values: KEYCODE_0,KEYCODE_1,KEYCODE_11,KEYCODE_12,KEYCODE_2,KEYCODE_3,KEYCODE_3D_MODE,KEYCODE_4,KEYCODE_5,KEYCODE_6,KEYCODE_7,KEYCODE_8,KEYCODE_9,KEYCODE_A,KEYCODE_ALL_APPS,KEYCODE_ALT_LEFT,KEYCODE_ALT_RIGHT,KEYCODE_APOSTROPHE,KEYCODE_APP_SWITCH,KEYCODE_ASSIST,KEYCODE_AT,KEYCODE_AVR_INPUT,KEYCODE_AVR_POWER,KEYCODE_B,KEYCODE_BACK,K ...
```
This renders all into one long line in my markdown viewer. Maybe it's better to put these in a list below the table?

fwolter reviewed (approved):
LGTM

cpmeister commented on bundles/org.openhab.binding.androiddebugbridge/pom.xml:
```diff
@@ -0,0 +1,25 @@
+    <dependency>
+      <groupId>com.tananaev</groupId>
+      <artifactId>adblib</artifactId>
+      <version>1.2</version>
+    </dependency>
```
In order for the dependency to be embedded into the binding jar you need to mark it with the compile scope.
```suggestion
      <version>1.2</version>
      <scope>compile</scope>
    </dependency>
```

cpmeister commented on bundles/org.openhab.binding.androiddebugbridge/NOTICE:
```diff
@@ -0,0 +1,13 @@
+
+== Source Code
+
+https://github.com/openhab/openhab-addons
```
Please add the licenses of your dependencies to this file.

cpmeister commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeDevice.java:
```diff
@@ -0,0 +1,299 @@
+
+    private String ip = "127.0.0.1";
+    private int port = 5555;
+    @Nullable
+    private Socket socket;
+    @Nullable
+    private AdbConnection connection;
```
```suggestion
    private @Nullable Socket socket;
    private @Nullable AdbConnection connection;
```

cpmeister commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeDiscoveryService.java:
```diff
@@ -0,0 +1,185 @@
+                                        if (message != null && message.contains("rejected by remote peer")) {
+                                            retries++;
+                                            logger.debug("retrying - pending {}", MAX_RETRIES - retries);
+                                            Thread.sleep(5000);
```
Please do not perform sleep operations in the `startScan` method. Those are called from a UI thread and as such you should make sure that `startScan` returns quickly.

GiviMAD replied:
That sleep operation is placed there to allow the user to confirm the host on their device  before a retry and avoid to the user to rerun the entire scan. Either way I can tell that the scan is quite slow because it checks every ip with in the defined range. 
Is it maybe better that I implement the discovery in background? 
I'm not sure how to do it, but maybe you can point me to a binding which uses background discovery.
Let me know your thoughts, if you think that just remove the sleep call is the best option I can go with that.

cpmeister commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeHandler.java:
```diff
@@ -0,0 +1,230 @@
+
+    @Override
+    public void dispose() {
+        super.dispose();
+        var schedule = connectionCheckerSchedule;
+        if (schedule != null) {
+            schedule.cancel(true);
+            connectionCheckerSchedule = null;
+        }
+        adbConnection.disconnect();
```
`super.dispose()` should be called last.

cpmeister commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeHandler.java:
```diff
@@ -0,0 +1,230 @@
+    }
+
+    private void refreshStatus() {
+        try {
+            handleCommandInternal(new ChannelUID(this.thing.getUID(), MEDIA_VOLUME_CHANNEL), RefreshType.REFRESH);
+        } catch (IOException | InterruptedException | AndroidDebugBridgeDevice.AndroidDebugBridgeDeviceException
+                | AndroidDebugBridgeDevice.AndroidDebugBridgeDeviceReadException e) {
+            logger.warn("Unable to refresh media volume: {}", e.getMessage());
+        }
+        try {
+            handleCommandInternal(new ChannelUID(this.thing.getUID(), MEDIA_CONTROL_CHANNEL), RefreshType.REFRESH);
+        } catch (IOException | InterruptedException | AndroidDebugBridgeDevice.AndroidDebugBridgeDeviceException
+                | AndroidDebugBridgeDevice.AndroidDebugBridgeDeviceReadException e) {
+            logger.warn("Unable to refresh play status: {}", e.getMessage());
+        }
+        try {
+            handleCommandInternal(new ChannelUID(this.thing.getUID(), CURRENT_PACKAGE_CHANNEL), RefreshType.REFRESH);
+        } catch (IOException | InterruptedException | AndroidDebugBridgeDevice.AndroidDebugBridgeDeviceException
+                | AndroidDebugBridgeDevice.AndroidDebugBridgeDeviceReadException e) {
+            logger.warn("Unable to refresh current package: {}", e.getMessage());
+        }
```
InterruptExceptions should cause an immediate return from the current method. You should not perform additional operations once an interrupt has been received since you would be delaying openhab from shutting down.

cpmeister commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeHandler.java:
```diff
@@ -0,0 +1,230 @@
+            handleCommandInternal(new ChannelUID(this.thing.getUID(), MEDIA_VOLUME_CHANNEL), RefreshType.REFRESH);
+        } catch (IOException | InterruptedException | AndroidDebugBridgeDevice.AndroidDebugBridgeDeviceException
+                | AndroidDebugBridgeDevice.AndroidDebugBridgeDeviceReadException e) {
+            logger.warn("Unable to refresh media volume: {}", e.getMessage());
```
Should such errors also cause the thing status to go offline?

GiviMAD replied:
It will allow to detect that the device is offline sooner, so yes I'll change it

cweitkamp replied:
For my Fire TV Stick this warning is logged all the time. I think we should reduce logging level here. Otherwise it will be very annoying. Or I have to reduce logging to ERROR for this binding in my logging config.

cpmeister commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeDiscoveryService.java:
```diff
@@ -0,0 +1,185 @@
+                            logger.debug("address: {}", currentIp);
+                            if (currentAddress.isReachable(discoveryReachableMs)) {
+                                logger.debug("Reachable ip: {}", currentIp);
+                                var retries = 0;
```
Please don't use `var` for primitives, it just leads to possible confusion.
```suggestion
                                int retries = 0;
```

cpmeister commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeDevice.java:
```diff
@@ -0,0 +1,299 @@
+        }
+    }
+
+    static class AndroidDebugBridgeDeviceException extends Exception {
+        public AndroidDebugBridgeDeviceException(String message) {
+            super(message);
+        }
+    }
+
+    static class AndroidDebugBridgeDeviceReadException extends Exception {
+        public AndroidDebugBridgeDeviceReadException(String message) {
+            super(message);
+        }
+    }
```
Since these exceptions are used outside of this class, please move these exceptions to their own files.

cpmeister commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeDiscoveryService.java:
```diff
@@ -0,0 +1,185 @@
+            for (NetworkInterface netint : Collections.list(nets)) {
+                Enumeration<InetAddress> inetAddresses = netint.getInetAddresses();
+                for (InetAddress inetAddress : Collections.list(inetAddresses)) {
+                    if (!discoveryRunning)
+                        break;
```
Please use brackets for every `if` statement.
```suggestion
                    if (!discoveryRunning) {
                        break;
                    }
```

GiviMAD commented:
I think everything is ready. Let me know if there something more I should improve and thanks for the previous help.

cweitkamp commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeDevice.java:
```diff
@@ -0,0 +1,285 @@
+        String audioDump = runAdbShell("dumpsys", "audio", "|", "grep", "ID:");
+        if (audioDump.length() == 0) {
+            // Fallback to media session dump
+            String devicesResp = runAdbShell("dumpsys", "media_session", "|", "grep", "PlaybackState");
```
I personally would check on "media_sesssion" first before "audio". At least for Android TV or Fire TV. Wdyt?

In general I am not very happy with the `isPlaying()` detection. E.g. My Fire TV Stick is playing all the time, event when I am on the home screen. Maybe we can add a configuration parameter to exclude specific, user-defined packages (e.g. "com.amazon.tv.launcher") for.

IIRC I somewhere read about kind of "state_detection_rules" for Android TV:

```
state_detection_rules = {
    'com.amazon.tv.launcher': ['idle'],
    'com.netflix.ninja': ['media_session_state'],
    'com.ellation.vrv': ['audio_state'],
    'com.hulu.plus': [{'playing': {'wake_lock_size' : 4}},{'paused': {'wake_lock_size': 2}}],
    'com.plexapp.android': [{'paused': {'media_session_state˓→': 3, 'wake_lock_size': 1}},{'playing': {'media_session_state˓→': 3}},'idle']
}
```

```
VALID_STATES = ('idle', 'off', 'playing', 'paused', 'standby')
```

GiviMAD replied:
I wasn't taking in account the current app for the playing check, now it should work better. 
Also I ended up implementing different ways of detect the player state, I have added the documentation. It's something similar to the solution implemented by HA. Basically you can configure a different detection mode by package name.

cweitkamp replied:
Did you test them? What is your personal feeling?

cweitkamp reviewed (commented):
Thanks for this binding. I compiled a test version and tested it for some days. Here are some thoughts.

GiviMAD commented:
I have also changed a few things in this last commit:
* I added a configurable timeout for the adb commands.
* I changed the way I read the binding configuration in the discovery to fix an exception (default values were not grabbed with the previous implementation).
* I added two new channels to allow read the wake lock state and the screen power.

Hope everything is in place this time.
Sorry for the late changes and thanks a lot for all the help.

cweitkamp commented:
I installed a version containing your latest changes and will check on the "state_dectection_rules". Thanks.

On think which came up after a power loss of the Fire TV is a recurring warning in my logs. I am not sure why, but it looks like the binding cannot connect anymore.

```
2021-01-09 14:21:03.978 [WARN ] [mmon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception: 
java.lang.IllegalStateException: connect() must be called first
	at com.tananaev.adblib.AdbConnection.open(AdbConnection.java:321) ~[?:?]
	at org.openhab.binding.androiddebugbridge.internal.AndroidDebugBridgeDevice.lambda$0(AndroidDebugBridgeDevice.java:267) ~[?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
2021-01-09 14:21:03.986 [WARN ] [mmon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception: 
java.lang.IllegalStateException: connect() must be called first
	at com.tananaev.adblib.AdbConnection.open(AdbConnection.java:321) ~[?:?]
	at org.openhab.binding.androiddebugbridge.internal.AndroidDebugBridgeDevice.lambda$0(AndroidDebugBridgeDevice.java:267) ~[?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
2021-01-09 14:21:03.994 [WARN ] [mmon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception: 
java.lang.IllegalStateException: connect() must be called first
	at com.tananaev.adblib.AdbConnection.open(AdbConnection.java:321) ~[?:?]
	at org.openhab.binding.androiddebugbridge.internal.AndroidDebugBridgeDevice.lambda$0(AndroidDebugBridgeDevice.java:267) ~[?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
2021-01-09 14:21:04.002 [WARN ] [mmon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception: 
java.lang.IllegalStateException: connect() must be called first
	at com.tananaev.adblib.AdbConnection.open(AdbConnection.java:321) ~[?:?]
	at org.openhab.binding.androiddebugbridge.internal.AndroidDebugBridgeDevice.lambda$0(AndroidDebugBridgeDevice.java:267) ~[?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
```

GiviMAD commented:
Now the thing connect again correctly. Thanks for pointing that.

cweitkamp commented:
Sry, I did non come back to you earlier. The latest version from last Sunday looks better. But can still see a warning in my logs when the binding tries to reconnect. I this situation my Fire TV Stick is in standby mode. Unfortunately it does not help to wake it up. I have to perform a full restart of the Fire TV Stick to establish aworking connection again.

```
2021-01-17 15:59:20.441 [WARN ] [mmon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception: 
java.lang.IllegalStateException: connect() must be called first
	at com.tananaev.adblib.AdbConnection.open(AdbConnection.java:321) ~[?:?]
	at org.openhab.binding.androiddebugbridge.internal.AndroidDebugBridgeDevice.lambda$0(AndroidDebugBridgeDevice.java:271) ~[?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
```

cweitkamp reviewed (commented):
some minor improvements

GiviMAD commented on bundles/org.openhab.binding.androiddebugbridge/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,403 @@
+			<channel id="media-volume" typeId="media-volume-channel"/>
+			<channel id="media-control" typeId="media-control-channel"/>
+			<channel id="start-package" typeId="start-package-channel"/>
+			<channel id="stop-package" typeId="stop-package-channel"/>
```
That channel allows to close an application by package name either if it is running in the background or not, I would like to keep like this. I'm using it in a rule right now to close the previous application when I send a new one, there I can add extra logic to avoid close the launcher or certain applications.
I will add another channel stop-current-package that works as you mention because you are right it could be more useful and simple in many situations.

GiviMAD commented:
I can't get rid of the warning in the logs because the library close the connection once the fire stick is offline and I not able to detect this until it throws that exception. Once it throws the binding should change the thing state to offline and the warning should stop spamming. 
I'll try to make a pr to the library to improve this in a future update but at the moment I couldn't get anything better or at least did manage to do it.
Let me know if more improvements are required and thanks for the feedback.

cweitkamp reviewed (approved):
Thanks for your patience. I am fine with the current implementation. Improvements and new features can be added later.

Only two final comments because I recognized it in the latest commit. First one is inline. Second one - please review the code of the `AndroidDebugBridgeHandler` class because there are some `if`-statements without curly brackets.

I am done. 😉

cweitkamp reviewed (approved):
LGTM. Thanks.

@cpmeister Are all your comments addressed?

cpmeister commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeHandlerFactory.java:
```diff
@@ -0,0 +1,52 @@
+@NonNullByDefault
+@Component(configurationPid = BINDING_CONFIGURATION_PID, service = ThingHandlerFactory.class)
+public class AndroidDebugBridgeHandlerFactory extends BaseThingHandlerFactory {
+    Gson gson = new Gson();
```
Instead of defining this here, define it in the AndroidDebugBridgeHandler as a static final constant.

GiviMAD commented on bundles/org.openhab.binding.androiddebugbridge/src/main/java/org/openhab/binding/androiddebugbridge/internal/AndroidDebugBridgeHandler.java:
```diff
@@ -0,0 +1,340 @@
+            this.packageConfigs = gson.fromJson(mediaStateJSONConfig,
+                    AndroidDebugBridgeMediaStatePackageConfig[].class);
+        } catch (JsonSyntaxException e) {
+            logger.warn("unable to parse media state config");
```
Yes, it should always return quickly, I have reviewed the implementation and I don't think its execution could be blocked by a network problem
